src/components/mgt-teams-channel-picker/mgt-teams-channel-picker.ts

```diff
@@ -0,0 +1,786 @@
+      this.selectChannel(item);
+    }
+
+    this._focusedIndex = -1;
+    this.resetFocusState();
+  }
+
+  private handleInputChanged(e) {
+    this._inputValue = e.target.value;
+    // shows list
+    this.gainedFocus();
+
+    if (!this.debouncedSearch) {
+      this.debouncedSearch = debounce(() => {
+        this.filterList();
+      }, 400);
+    }
+
+    this.debouncedSearch();
+  }
+
+  private filterList() {
+    if (this.items) {
+      this._treeViewState = this.generateTreeViewState(this.items, this._inputValue);
+      this._focusedIndex = -1;
+      this.resetFocusState();
+    }
+  }
+
+  private generateTreeViewState(
+    tree: DropdownItem[],
+    filterString: string = '',
+    parent: ChannelPickerItemState = null
+  ): ChannelPickerItemState[] {
+    const treeView: ChannelPickerItemState[] = [];
+    filterString = filterString.toLowerCase();
+
+    if (tree) {
+      for (const state of tree) {
+        let stateItem: ChannelPickerItemState;
+
+        if (filterString.length === 0 || state.item.displayName.toLowerCase().includes(filterString)) {
+          stateItem = { item: state.item, parent };
+          if (state.channels) {
+            stateItem.channels = this.generateTreeViewState(state.channels, '', stateItem);
+            stateItem.isExpanded = filterString.length > 0;
+          }
+        } else if (state.channels) {
+          const newStateItem = { item: state.item, parent };
+          const channels = this.generateTreeViewState(state.channels, filterString, newStateItem);
+          if (channels.length > 0) {
+            stateItem = newStateItem;
+            stateItem.channels = channels;
+            stateItem.isExpanded = true;
+          }
+        }
+
+        if (stateItem) {
+          treeView.push(stateItem);
+        }
+      }
+    }
+    return treeView;
+  }
+
+  // generates a flat list from a tree to facilitate easier focus
+  // navigation
+  private generateFocusList(items): any[] {
+    if (!items || items.length === 0) {
+      return [];
+    }
+
+    let array = [];
+
+    for (const item of items) {
+      array.push(item);
+      if (item.channels && item.isExpanded) {
+        array = [...array, ...this.generateFocusList(item.channels)];
+      }
+    }
+
+    return array;
+  }
+
+  private resetFocusState() {
+    this._focusList = this.generateFocusList(this._treeViewState);
+    this.requestUpdate();
+  }
+
+  private loadTeamsIfNotLoaded() {
+    if (!this.items && !this.isLoadingState) {
+      this.requestStateUpdate();
+    }
+  }
+
+  private handleWindowClick(e: MouseEvent) {
+    if (e.target !== this) {
+      this.lostFocus();
+    }
+  }
+
+  private onUserKeyDown(event: KeyboardEvent) {
+    if (this._treeViewState.length === 0) {
+      return;
+    }
+
+    const currentFocusedItem = this._focusList[this._focusedIndex];
+
+    switch (event.keyCode) {
+      case 40: // up
+        this._focusedIndex = (this._focusedIndex + 1) % this._focusList.length;
```

Comment:
A few comments:

1. `// up` and `// down` are swapped. Pressing up hits case 38, down hits case 40.
1. When making a search, the first press up will set focus on the second to last item, instead of the very last item. This should fix it:
    ```
    case 38: // up
      if (this._focusedIndex === -1) {
        this._focusedIndex = this._focusList.length;
      }
      this._focusedIndex = (this._focusedIndex - 1 + this._focusList.length) % this._focusList.length;
      this.requestUpdate();
      event.preventDefault();
      break;
    ```
1. I see these number type checks in both up/down cases. When does this happen?
    ```
    if (typeof this._focusedIndex !== 'number') {
      this._focusedIndex = -1;
    }
    ```

Reply:
Thanks for the through review! I've updated the comments and the up search, and it works well 👍 

As for the number type checks, maybe @nmetulev can speak to this? I can't see a reason that its necessary in either case, but it might be leftover logic.

Reply:
unless `this._focusIndex` is set to null somewhere, I don't see this being necessary